### PR TITLE
Add mechanism for database migration

### LINKED ISSUE #77: Thoughts on database management/migrations
#69 made me think about how we want to handle database management in general. I researched and thought a bit about this topic. These are my thoughts/conclusions.

--- 

Some notes (hopefully there is no disagreement on those?):

- Releasing a new version of Tobira should include some easy way to migrate the old database to the new schema. Otherwise updating is terribly annoying for sys admins.
- I want commands for DB operations, e.g. `tobira db setup`, `tobira db reset` and `tobira db run script.sql`.
- For convenience, just starting `tobira` should do some setup/migrations automatically, as long as it's clear what to do. In case of ambiguity, Tobira shouldn't guess.
- It will regularly happen that we (devs) switch between versions requiring different DB schemas (e.g. git checkout to review a PR, ...). It shouldn't be too annoying to get the database into the correct state. Notably, for development, the actual data in the database isn't important most of the time.
- I would prefer the scripts to be written in actual `.sql` files and not Rust strings. We can still embed those files into the application (`include_str!`).

## Suggestion: `down.sql`

In addition to the obviously needed "forwards" migration (`up.sql`), I think it would be beneficial to also include scripts to roll back one migration. 

## Suggestion: store migrations in DB

We would add a table `migrations` (or so) which would contain all migrations leading to the current state of the database. All migrations should have some unique ID. Running a migration would add a row to this table, undoing a migration would remove the row.

Furthermore, I think it would be useful for development to also store the raw `up.sql` and `down.sql` in a `text` column:

Saving `down.sql` allows us to roll back a migration even if the script is not currently available in the repository. That is useful if you switch from one feature branch (`foo`) to a different one (`bar`). To get the database in the state expected by `bar`, we have to roll back the migration of `foo` and then run the migation of `bar`. But since `foo`'s `down.sql` is not in the working directory anymore, we need to store it in the DB. Of course, you can always manually undo `foo`'s migration before switching to the new branch, but we will likely forget that most of the time and get annoyed by how inconvenient that is.

Saving `up.sql` allows us to compare the `up.sql` in the working directory/file system with the one in the database. If they differ (which will happen a lot during development), Tobira could easily undo the old migration (with the `down.sql` in the database) and then run the new `up.sql` from the file system.

### Solution A: one migration per version

One migration script per version (e.g. `v1.0.sql`, `v1.1.sql` and `next.sql`). We would only modify the latest (e.g. `next.sql`) during development. Releasing a new version means `mv next.sql 1.2.sql && touch next.sql`. The ID for the migrations in the `migration` table would simply be the version. That way, Tobira knows what migrations to run when updating to a new version.

- Advantage: works great for production users as they can easily upgrade to a new version
- Advantage: having all schema changes between two versions in one file helps on some situations
- Advantage: migration potentially faster as there are fewer scripts
- Disadvantage: for developers, switching between branches means to completely undo all database changes since the last released version of Tobira and then redoing them.

### Solution B: one migration per "feature" (commit/PR)

Instead of continuously changing one script until release, we could also have a separate migration script per "feature". That roughly corresponds to one migration script per PR that requires to modify the database schema. This of course results in lot more migration scripts, but also probably not ungodly many (adjust the DB schema is not too common). As ID for migrations, we would probably either want to use the current timestamp or some random short string. This is mainly to ensure that two feature branches developed in parallel will have different IDs for their migrations.

- Still works fine for production users, although it might be a bit slower than solution A, simply because there are more migrations to run.
- Advantage: more atomic migrations could make error handling/debugging easier
- Advantage: for development, switching between branches would only require undoing a small migration instead of basically resetting the whole database.
- Disadvantage: in solution (a) we could merge changes easily, whereas here, we might end up with migrations that undo part of previous migrations. Maybe. 

I at least know of one library which encourages this style of migration management.

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Closes #77 

See commit messages. 

**TODO** (tho maybe not in this PR?):
- [ ] Add `tobira db migrate [--up-to <index>]` subcommand to manually run the migrations
- [ ]  Think whether running migrations scripts on startup automatically is really a good idea. Then we really have to make sure that the scripts never destroy data in the DB. It is convenient tho.
- [ ] What about multiple tobira nodes? What if all of them start migrating at the same time? Does something break? How to basically globally lock the full DB when running migations?
- [ ] Add mechanism to have checks that can make sure certain things in the DB are correct.

### CODE REVIEW

**Comment by LukasKalbertodt** [comment]
This would be ready to merge now. I think it's fine if we don't include any of the optional tasks in the top comment in this PR. The open questions one should still think about before merging or create dedicated issues for.

**Comment by JulianKniephoff** [comment]
> What about multiple tobira nodes? What if all of them start migrating at the same time? Does something break? How to basically globally lock the full DB when running migations?

Just wrap the whole thing in a transaction? :thinking:

**Comment by LukasKalbertodt** [comment]
> > What about multiple tobira nodes? What if all of them start migrating at the same time? Does something break? How to basically globally lock the full DB when running migations?
> 
> Just wrap the whole thing in a transaction?

Does this work? Only if we check within the transaction whether the table already exists, right? Otherwise this might happen, right?

1. Node A checks (in Rust) whether `__db_migrations` exists. -> No
2. Node B checks (in Rust) whether `__db_migrations` exists. -> No
3. Node A & B prepare a long transaction with all migrations.
4. Node B is done and commits the transaction.
5. Node A is done and commits the transaction, but this will now run everything twice, because it is not re-checked whether the `__db_migrations` table exists, right?

**Comment by JulianKniephoff** [comment]
For the record: As we just discussed, with the right isolation level you are guaranteed that 1. and 2. never happen simultaneously while both of the transactions succeed.

**Review by JulianKniephoff — CHANGES_REQUESTED** [review]
Very cool! Some minor things, again, and just so that it is also mentioned "officially" in the review again: I would just wrap the whole migration process in one big transaction, using the correct isolation level. And that can probably easily be added as part of this PR, right?

We can then also think about how to handle errors. If the migration fails because of conflicts on concurrent transactions, we might have some kind of retry loop. Or at least inform the user about the fact that they will probably be fine if they just run it again. But this doesn't have to be in this PR, IMO.

File: backend/server/src/db/query.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+
+/// Returns meta information about all tables in the `public` schema.
+pub(super) async fn all_table_names(db: &Db) -> Result<Vec<String>> {
+    let rows = db.query_raw(
+            "select table_name from information_schema.tables where table_schema='public'",
+            std::iter::empty(),
+        )
+        .await?
+        .map_ok(|row| row.get::<_, String>("table_name"));
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply backend/server/src/db/query.rs:14]: Yeah good point. I mean, it's way less important for this code that doesn't run often. But yeah.

**Comment by LukasKalbertodt** [comment]
I think I fixed everything. Regarding the transaction: my brain does not yet fully accept that this works as expected but yeah, I guess you are right. We use a single serializable transaction now. And I also hope I correctly built the retry-loop. 

I have **not** tested this yet with two actual nodes! Testing this is a bit involved and annoying. I created #109 so that we don't forget that.

**Review by JulianKniephoff — CHANGES_REQUESTED** [review]
Ok cool, this looks all good now, I think. Except for one little thing: `scripts/schema.sql` can probably be deleted now, right?

**Review by JulianKniephoff — APPROVED** [review]

**Pull request merged**